test/unit/math/rev/prob/neg_binomial_2_log_test.cpp

```diff
@@ -0,0 +1,548 @@
+    TestValue(1525, 368, 0.065, -33.65022328528903, -0.065, -346.5441943953198),
+    TestValue(10233, 368, 0.065, -35.430094188966905, -0.065,
+              -344.6403282546427),
+    TestValue(0, 368, 4.42, -1619.991262543284, -4.420000000000001,
+              -365.5138603039104),
+    TestValue(6, 368, 4.42, -1615.1572361520293, -4.420000000000001,
+              -364.58765802750503),
+    TestValue(14, 368, 4.42, -1612.8136580987152, -4.420000000000001,
+              -363.99658652545247),
+    TestValue(1525, 368, 4.42, -1597.2619947735784, -4.420000000000001,
+              -359.55031671340777),
+    TestValue(10233, 368, 4.42, -1590.7558162463538, -4.420000000000001,
+              -357.6488775415163),
+    TestValue(0, 368, 800, -289052.31061786565, -800., -360.3153882723321),
+    TestValue(6, 368, 800, -289018.7634915345, -800., -360.30791160295655),
+    TestValue(14, 368, 800, -288983.8041593674, -800., -360.29802888017133),
+    TestValue(1525, 368, 800, -287560.8055295305, -800., -359.2481146209497),
+    TestValue(10233, 368, 800, -286189.66239505477, -800., -357.6907741283057),
+    TestValue(0, 368, 15324, -5.491551922548806e6, -15324.,
+              -357.36282449417945),
+    TestValue(6, 368, 15324, -5.4915006777682435e6, -15324.,
+              -357.36243301536337),
+    TestValue(14, 368, 15324, -5.491442187376253e6, -15324.,
+              -357.36191128189233),
+    TestValue(1525, 368, 15324, -5.486439246638204e6, -15324.,
+              -357.2679504598954),
+    TestValue(10233, 368, 15324, -5.474353427491954e6, -15324.,
+              -356.85132040461684),
+};
+
+}  // namespace neg_binomial_2_log_test_internal
+
+TEST(ProbDistributionsNegBinomial2Log, derivativesPrecomputed) {
+  using neg_binomial_2_log_test_internal::TestValue;
+  using neg_binomial_2_log_test_internal::testValues;
+  using stan::math::is_nan;
+  using stan::math::neg_binomial_2_log_lpmf;
+  using stan::math::value_of;
+  using stan::math::var;
+
+  for (TestValue t : testValues) {
+    int n = t.n;
+    var eta(t.eta);
+    var phi(t.phi);
+    var val = neg_binomial_2_log_lpmf(n, eta, phi);
+
+    std::vector<var> x;
+    x.push_back(eta);
+    x.push_back(phi);
+
+    std::vector<double> gradients;
+    val.grad(x, gradients);
+
+    for (int i = 0; i < 2; ++i) {
+      EXPECT_FALSE(is_nan(gradients[i]));
+    }
+
+    auto tolerance = [](double x) { return std::max(fabs(x * 1e-8), 1e-14); };
+
+    EXPECT_NEAR(value_of(val), t.value, tolerance(t.value))
+        << "value n = " << n << ", eta = " << t.eta << ", phi = " << t.phi;
+    EXPECT_NEAR(gradients[0], t.grad_eta, tolerance(t.grad_eta))
+        << "grad_mu n = " << n << ", eta = " << t.eta << ", phi = " << t.phi;
+    EXPECT_NEAR(gradients[1], t.grad_phi, tolerance(t.grad_phi))
+        << "grad_phi n = " << n << ", eta = " << t.eta << ", phi = " << t.phi;
+  }
+}
+
+TEST(ProbDistributionsNegBinomial2, derivativesComplexStep) {
```

Comment:
> What do the complex step tests cover that the others don't?

They are able to test gradients numerically to (much) higher precision than what is possible with finite diffs. They are also able to cover large `phi` values where Mathematica (at least the free cloud version) gives up on computing exact answers.

> These fancy tests are for neg_binomial_2_log_lpmf. Should they cover neg_binomial_2_lpmf as well?

Well, they should :-) And I do have them written in #1497 . The thing is I am aiming for really a minimal PR and the tests fail (unless weakened noticeably), because there are all sorts of minor issues with the numerics that require more time to be handled. I don't think there is any particularly defensible line on what tests to include (and hence what issues to address). I went with "avoid obvious bugs (positive log prob., overwriting `logp`)". 

You may disagree on the specific choice or with the whole concept of the quick fix, obviously. I am myself not 100% happy with leaving so many holes, and maybe waiting for a full fix is a better option in some sense.